File: packages/native/router-swift/Sources/PyreonRouter/RouteLoader.swift
// RouteLoader — the SwiftUI host the compiler wraps a loader-bearing
// route's component in. Mirrors @pyreon/router's per-route `loader`:
// the loader runs ONCE when the route's view appears, and stores its
// result on the active router's `loaderData[path]`, where the
// already-shipped `useLoaderData<T>()` reads it.
//
// Compiler emit (Phase 3 — per-route loaders):
//   { path: '/dashboard', component: Dashboard, loader: () => fetchStats() }
//   ↓ (inside the navigationDestination / RouterView dispatch)
//   PyreonRouteLoader(path: "/dashboard", load: { fetchStats() }) {
//       Dashboard()
//   }
//
// Lifecycle note (the fetch-arc lesson): `.task` ties its lifetime to
// its host VIEW's identity. Here it attaches to `content()` — the route's
// concrete component view, which has STABLE identity for the duration of
// the route render (it is NOT a transparent `Group { if … }`, so SwiftUI
// does not redistribute the modifier onto a flipping branch and cancel /
// restart it). So `.task` fires exactly once on appear. The
// `loaderData[path] == nil` guard makes the store idempotent across any
// re-evaluation, matching the web router's "loader runs once per nav".

import SwiftUI

/// Wraps a route's component, firing its `loader` once on appear and
/// storing the result via `router.setLoaderData(path, …)`.
///
/// v1 scope: the `load` closure is SYNCHRONOUS (the compiler emits a
/// zero-param, expression-body loader). Truly-async `await` bodies +
/// `ctx.params` threading are a later arc — see `RouteIR.loader`.
@available(iOS 17.0, macOS 14.0, *)
public struct PyreonRouteLoader<Content: View>: View {
    @Environment(\.pyreonRouter) private var router: PyreonRouter?
    private let path: String
    private let load: () -> Any
    private let content: () -> Content

    public init(
        path: String,
        load: @escaping () -> Any,
        @ViewBuilder content: @escaping () -> Content
    ) {
        self.path = path
        self.load = load
        self.content = content
    }

    public var body: some View {
        content().task {
            // Idempotent: only run the loader if this route has no stored
            // data yet (a re-appear / re-evaluation must not re-run it).
            if let router, router.loaderData[path] == nil {
                router.setLoaderData(path, load())
            }
        }
    }
}

File: packages/native/router-kotlin/src/main/kotlin/com/pyreon/router/RouteLoader.kt
// RouteLoader — the Compose host the compiler wraps a loader-bearing
// route's component in. Mirrors @pyreon/router's per-route `loader`:
// the loader runs ONCE when the route's content enters composition, and
// stores its result on the active router's `loaderData[path]`, where the
// already-shipped `useLoaderData<T>()` reads it.
//
// Compiler emit (Phase 3 — per-route loaders):
//   { path: '/dashboard', component: Dashboard, loader: () => fetchStats() }
//   ↓ (inside the when-dispatch / RouterView dispatch)
//   PyreonRouteLoader(path = "/dashboard", load = { fetchStats() }) {
//       Dashboard()
//   }
//
// Lifecycle note: `LaunchedEffect(path)` runs once when the content
// enters composition (and again only if `path` changes — which it never
// does for a given dispatch branch). Unlike SwiftUI's `.task`, a
// `LaunchedEffect` keyed by a stable value is NOT cancelled/restarted by
// recomposition, so no stable-host wrapping is needed (the Kotlin
// counterpart of the fetch-arc ZStack concern). The
// `loaderData.value[path] == null` guard makes the store idempotent,
// matching the web router's "loader runs once per nav".

package com.pyreon.router

import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect

/**
 * Wraps a route's component, firing its `loader` once on entering
 * composition and storing the result via `router.setLoaderData(path, …)`.
 *
 * v1 scope: the [load] lambda is SYNCHRONOUS (the compiler emits a
 * zero-param, expression-body loader). Truly-async (`suspend` / `await`)
 * bodies + `ctx.params` threading are a later arc — see `RouteIR.loader`.
 */
@Composable
public fun PyreonRouteLoader(
    path: String,
    load: () -> Any?,
    content: @Composable () -> Unit,
) {
    val router = LocalPyreonRouter.current
    LaunchedEffect(path) {
        // Idempotent: only run the loader if this route has no stored
        // data yet (a recomposition must not re-run it).
        if (router != null && router.loaderData.value[path] == null) {
            router.setLoaderData(path, load())
        }
    }
    content()
}
